src/api.js

```diff
@@ -464,56 +511,104 @@ var remoteAuthentication = function(options, entry) {
           return false;
         }
 
-        // If we're not given an array, we assume it's a set of parameters that
-        // must be used to parameterize the original scopesets
-        if (!(scopesets instanceof Array)) {
-          var params = scopesets;
-          scopesets = _.cloneDeepWith(entry.scopes, function(scope) {
+        req.hasAuthed = false;
+        let missing = [];
+        let hasConditional = false;
+
+        var _splatParams = (scope) => scope.replace(/<([^>]+)>/g, function(match, param) {
+          var value = _.at(params, param)[0];
+          if (value !== undefined) {
+            return value;
+          }
+          missing.push(match); // If any are left undefined, we can't be done yet
+          return match;
+        });
+
+        var _expandExpressionTemplate = (template) => {
+          var key = Object.keys(template)[0];
+          var subexpressions = [];
+          template[key].forEach(scope => {
             if (typeof scope === 'string') {
-              return scope.replace(/<([^>]+)>/g, function(match, param) {
-                var value = params[param];
-                return value === undefined ? match : value;
+              subexpressions.push(_splatParams(scope));
+            } else if (_.isObject(scope) && scope.for && scope.in && scope.each) {
+              let subs = _.at(params, scope.in)[0];
+              if (!subs) {
+                missing.push(scope.in);
+                subexpressions.push(scope);
+                return;
+              }
+              subs.forEach(param => {
+                subexpressions.push(_splatParams(scope.each.replace(`<${scope.for}>`, param)));
               });
+            } else if (_.isObject(scope) && scope.if && scope.then) {
+              hasConditional = true;
+              if (_.at(params, scope.if)[0]) {
+                subexpressions.push(_expandExpressionTemplate(scope.then));
+              }
+            } else {
+              subexpressions.push(_expandExpressionTemplate(scope));
             }
           });
+          return {[key]: subexpressions};
+        };
+
+        var scopeExpression = _expandExpressionTemplate(entry.scopes);
+
+        if (hasConditional && allowLater) {
+          debug('Conditional provided in scopes. Deferring authorization.');
```

Comment:
Why is this the case?  If the value for the conditional wasn't missing, isn't it fine?

Reply:
The general issue here is that if the conditional exists, we don't know that we are allowed to ignore results from the initial authorize that happens automatically. With parameters, we know that we can skip authorizing if any are missing, but with `if/then` we can't really know. 

An example is `scopes: {AnyOf: ['whatever', {if: 'public', then: {AllOf: []}}]}`. When we initially call the function, we don't know `public` yet, so this evaluates to `{AnyOf: ['whatever']}` which will fail for public requests with no scopes. We will never get to the part where we call `authorize` later in this case.

Now, obviously we can fix this specific case by doing jonas' more clever  `{AllOf: [{if: 'not_public', then: 'foo:bar'}]}`, but now we're at the point where we're requiring people to craft scope expressions in such a way that they must always succeed when initially evaluated  by the endpoint in order to allow them to call `authorize` explicitly later. Really what we could use here is adding back in a `deferAuth` flag I guess? I'm sorta not loving the `if/then` construct we have. It complicates things a good deal.

Reply:
We figured this out in irc!